src/integrator.jl

```diff
@@ -5,37 +5,76 @@
 
 abstract type AbstractIntegrator end
 
-struct Leapfrog{T<:AbstractFloat} <: AbstractIntegrator
-    ϵ       ::  T
-    jitter  :: T
-end
-Base.show(io::IO, l::Leapfrog) = print(io, "Leapfrog(ϵ=$(round(l.ϵ; sigdigits=3)))")
+abstract type AbstractLeapfrog{T} <: AbstractIntegrator end
 
-Leapfrog(ϵ::T) where {T<:AbstractFloat} = Leapfrog(ϵ, zero(T))
+jitter(::AbstractLeapfrog, ϵ) = ϵ
+temper_first(::AbstractLeapfrog, r, ::Int, ::Int) = r
+temper_second(::AbstractLeapfrog, r, ::Int, ::Int) = r
 
-# TODO: double check the function below to see if it is type stable or not
 function step(
-    lf::Leapfrog{F},
+    lf::AbstractLeapfrog{T},
     h::Hamiltonian,
     z::PhasePoint,
     n_steps::Int=1;
-    fwd::Bool=n_steps > 0 # simulate hamiltonian backward when n_steps < 0,
-) where {F<:AbstractFloat,T<:Real}
-    @unpack θ, r = z
+    fwd::Bool=n_steps > 0   # simulate hamiltonian backward when n_steps < 0,
+) where {T<:AbstractFloat}
+    n_steps = abs(n_steps)  # to support `n_steps < 0` cases
     ϵ = fwd ? lf.ϵ : -lf.ϵ
-    # Jitter step size; ref: https://github.com/stan-dev/stan/blob/1bb054027b01326e66ec610e95ef9b2a60aa6bec/src/stan/mcmc/hmc/base_hmc.hpp#L177-L178
-    iszero(lf.jitter) || (ϵ *= (1 + lf.jitter * (2 * rand() - 1)))
+    ϵ = jitter(lf, ϵ)
 
+    @unpack θ, r = z
     @unpack value, gradient = ∂H∂θ(h, θ)
-    for i = 1:abs(n_steps)
-        r = r - ϵ / 2 * gradient # take a half leapfrog step for momentum variable
+    for i = 1:n_steps
+        r = temper_first(lf, r, i, n_steps)
+        r = r - ϵ / 2 * gradient    # take a half leapfrog step for momentum variable
         ∇r = ∂H∂r(h, r)
-        θ = θ + ϵ * ∇r   # take a full leapfrog step for position variable
+        θ = θ + ϵ * ∇r              # take a full leapfrog step for position variable
         @unpack value, gradient = ∂H∂θ(h, θ)
-        r = r - ϵ / 2 * gradient # take a half leapfrog step for momentum variable
+        r = r - ϵ / 2 * gradient    # take a half leapfrog step for momentum variable
+        r = temper_second(lf, r, i, n_steps)
         # Create a new phase point by caching the logdensity and gradient
         z = phasepoint(h, θ, r; ℓπ=DualValue(value, gradient))
         !isfinite(z) && break
     end
     return z
 end
+
+struct Leapfrog{T<:AbstractFloat} <: AbstractLeapfrog{T}
+    ϵ       ::  T
+end
+Base.show(io::IO, l::Leapfrog) = print(io, "Leapfrog(ϵ=$(round(l.ϵ; sigdigits=3)))")
+
+### Jittering
+
+struct JitteredLeapfrog{T<:AbstractFloat} <: AbstractLeapfrog{T}
+    ϵ       ::  T
+    jitter  ::  T
+end
+
+function Base.show(io::IO, l::JitteredLeapfrog)
+    print(io, "JitteredLeapfrog(ϵ=$(round(l.ϵ; sigdigits=3)), jitter=$(round(l.jitter; sigdigits=3)))")
+end
+
+# Jitter step size; ref: https://github.com/stan-dev/stan/blob/1bb054027b01326e66ec610e95ef9b2a60aa6bec/src/stan/mcmc/hmc/base_hmc.hpp#L177-L178
+jitter(lf::JitteredLeapfrog, ϵ) = ϵ * (1 + lf.jitter * (2 * rand() - 1))
```

Comment:
Consider adding a `RNG` argument here since it relies on random generator.

Reply:
Good point!

Reply:
The currently `step` doesn't pass RNG so there are two options here.
1. Pass RNG to `step`
2. Add RNG to `JitteredLeapfrog`
How do you think?

Reply:
Sounds good!